py/picca/io.py

```diff
@@ -261,42 +260,114 @@ def read_data(in_dir,drq,mode,zmin = 2.1,zmax = 3.5,nspec=None,log=None,keep_bal
 
     return data,ndata,nside,"RING"
 
-def read_from_spec(in_dir,thid,ra,dec,zqso,plate,mjd,fid,order,mode,log=None,pk1d=None):
+def read_from_spec(in_dir,thid,ra,dec,zqso,plate,mjd,fid,order,mode,log=None,pk1d=None,best_obs=None):
+    drq_dict = {t:(r,d,z) for t,r,d,z in zip(thid,ra,dec,zqso)}
+
+    ## if using multiple observations,
+    ## then replace thid, plate, mjd, fid
+    ## by what's available in spAll
+    if not best_obs:
+        fi = glob.glob(in_dir.replace("spectra/","").replace("lite","").replace("full","")+"/spAll-*.fits")
+        if len(fi) > 1:
+            print("ERROR: found multiple spAll files")
+            print("ERROR: try running with --bestobs option (but you will lose reobservations)")
+            for f in fi:
+                print("found: ",fi)
+            sys.exit(1)
+        if len(fi) == 0:
+            print("ERROR: can't find required spAll file in {}".format(in_dir))
+            print("ERROR: try runnint with --best-obs option (but you will lose reobservations)")
+            sys.exit(1)
+
+        spAll = fitsio.FITS(fi[0])
+        print("INFO: reading spAll from {}".format(fi[0]))
+        thid_spall = spAll[1]["THING_ID"][:]
+        plate_spall = spAll[1]["PLATE"][:]
+        mjd_spall = spAll[1]["MJD"][:]
+        fid_spall = spAll[1]["FIBERID"][:]
+        qual_spall = spAll[1]["PLATEQUALITY"][:]
+        zwarn_spall = spAll[1]["ZWARNING"][:]
+
+        w = sp.in1d(thid_spall, thid) & (qual_spall == b"good")
+        ## Removing spectra with the following ZWARNING bits set:
+        ## SKY, LITTLE_COVERAGE, UNPLUGGED, BAD_TARGET, NODATA
+        ## https://www.sdss.org/dr14/algorithms/bitmasks/#ZWARNING
+        for zwarnbit in [0,1,7,8,9]:
+            w &= (zwarn_spall&2**zwarnbit)==0
+        print("INFO: # unique objs: ",len(thid))
+        print("INFO: # spectra: ",w.sum())
+        thid = thid_spall[w]
+        plate = plate_spall[w]
+        mjd = mjd_spall[w]
+        fid = fid_spall[w]
+        spAll.close()
+
+    ## to simplify, use a list of all metadata
+    allmeta = []
+    ## Used to preserve original order and pass unit tests.
+    t_list = []
+    t_set = set()
```

Comment:
this should work in principle, but down there you still iterate over t_list instead of t_set.
I think it's probably more readable if you just make `thids` an `collections.OrderedDict` and use `items()` or `keys()`

Reply:
Sorry about the delay. I was thinking about that other data structure but I feel like it's better to not import libraries and just stick with the default Python data structures. I can use it if you would prefer. The point of using the set was to keep track of what thids were added to the list and prevent duplicates from being added. Sets have constant time lookup so I felt like the performance overhead would be minimal.

Reply:
The set is fine with me, but you still need to iterate over the set in the for loop of line 338.

Reply:
I use the list to preserve order which the set doesn't. The set is just temporarily used to prevent duplicates thids from being added to the list.